dataplane/src/nat/fabric.rs

```diff
@@ -0,0 +1,419 @@
+    }
+
+    #[tracing::instrument(level = "trace")]
+    pub fn find_prefix(&self, ip: &IpAddr) -> Option<&Prefix> {
+        self.iter_endpoints().find(|&prefix| prefix.covers_addr(ip))
+    }
+}
+
+#[derive(Debug, Clone, Deserialize)]
+pub struct PeeringPolicy {
+    name: String,
+    vnis: [Vni; 2],
+    pifs: [String; 2],
+}
+
+impl PeeringPolicy {
+    #[tracing::instrument(level = "trace")]
+    pub fn new(name: String, vnis: [Vni; 2], pifs: [String; 2]) -> Self {
+        Self { name, vnis, pifs }
+    }
+
+    #[tracing::instrument(level = "trace")]
+    pub fn name(&self) -> &String {
+        &self.name
+    }
+
+    #[tracing::instrument(level = "trace")]
+    pub fn vnis(&self) -> &[Vni; 2] {
+        &self.vnis
+    }
+
+    #[tracing::instrument(level = "trace")]
+    pub fn pifs(&self) -> &[String; 2] {
+        &self.pifs
+    }
+
+    #[tracing::instrument(level = "trace")]
+    pub fn get_index(&self, pif: &Pif) -> usize {
+        usize::from(self.pifs[0] != pif.name)
+    }
+
+    #[tracing::instrument(level = "trace")]
+    pub fn get_peer_index(&self, pif: &Pif) -> usize {
+        return self.get_index(pif) ^ 1;
+    }
+
+    #[tracing::instrument(level = "trace")]
+    pub fn get_peer(&self, pif: &Pif) -> &String {
+        &self.pifs[self.get_peer_index(pif)]
+    }
+}
+
+#[derive(Debug, Clone, Deserialize, Serialize)]
+pub struct Vpc {
+    name: String,
+    vni: Vni,
+    pif_table: PifTable,
+}
+
+impl Vpc {
+    #[tracing::instrument(level = "trace")]
+    pub fn new(name: String, vni: Vni) -> Self {
+        Self {
+            name,
+            vni,
+            pif_table: PifTable::new(),
+        }
+    }
+
+    #[tracing::instrument(level = "trace")]
+    pub fn name(&self) -> &String {
+        &self.name
+    }
+
+    #[tracing::instrument(level = "trace")]
+    pub fn vni(&self) -> Vni {
+        self.vni
+    }
+
+    #[tracing::instrument(level = "trace")]
+    pub fn add_pif(&mut self, pif: Pif) -> Result<(), TrieError> {
+        self.pif_table.add_pif(pif)
+    }
+
+    #[tracing::instrument(level = "trace")]
+    pub fn find_pif_by_endpoint(&self, ip: &IpAddr) -> Option<String> {
+        self.pif_table.find_pif_by_endpoint(ip)
+    }
+
+    #[tracing::instrument(level = "trace")]
+    pub fn get_pif(&self, name: &String) -> Option<&Pif> {
+        self.pif_table.pifs.get(name)
+    }
+
+    #[tracing::instrument(level = "trace")]
+    pub fn iter_pifs(&self) -> impl Iterator<Item = &Pif> {
+        self.pif_table.pifs.values()
+    }
+}
+
+#[derive(Debug, Default, Clone)]
+struct PifTable {
+    pifs: HashMap<String, Pif>,
+    endpoint_trie: PrefixTrie,
+}
+
+impl PifTable {
+    #[tracing::instrument(level = "trace")]
+    fn new() -> Self {
+        Self {
+            pifs: HashMap::new(),
+            endpoint_trie: PrefixTrie::new(),
+        }
+    }
+
+    #[tracing::instrument(level = "info")]
+    fn add_pif(&mut self, pif: Pif) -> Result<(), TrieError> {
```

Comment:
If Piftable is a global table containing all of the PIFs, I believe that the key should be (vpc, pifname), right? Otherwise, there may be name collisions. IMO, each VPC should be allowed to choose the pif names.

Reply:
Oh, now I see that pif table is per VPC. Ignore my comment!